runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -348,16 +348,33 @@ public void prefixTrim(final Token address) {
                                     .flatMap(stripe -> stripe.getLogServers().stream())
                                     .map(e::getLogUnitClient)
                                     .map(client -> client.prefixTrim(address))
-                                    .forEach(CFUtils::getUninterruptibly);
-                            return null;    // No return value
-                        }, true);
+                                    .forEach(cf -> {CFUtils.getUninterruptibly(cf,
+                                            NetworkException.class, TimeoutException.class,
+                                            WrongEpochException.class);
+                                    });
+                            return null;
+                }, true);
                 // TODO(Maithem): trimCache should be epoch aware?
                 runtime.getSequencerView().trimCache(address.getSequence());
                 break;
             } catch (NetworkException | TimeoutException e) {
                 log.warn("prefixTrim: encountered a network error on try {}", x, e);
                 Duration retryRate = runtime.getParameters().getConnectionRetryRate();
                 Sleep.sleepUninterruptibly(retryRate);
+            } catch (WrongEpochException wee) {
+                log.warn("prefixTrim[{}]: wrongEpochException", address);
+                long runtimeEpoch = runtime.getLayoutView().getLayout().getEpoch();
+                long serverEpoch = wee.getCorrectEpoch();
+                // Only retry if wrongEpochException corresponds to message epoch
+                if (runtimeEpoch != serverEpoch) {
```

Comment:
It seems like a wrong epoch exception can originate from the server router, or the batch writer. 
The batch writer check the epoch of address.getEpoch, bit the server epoch will check the msg.getEpoch, which is stamped by the client's layout. So why not only retry if address.getEpoch == serverEpoch? 

Because the current check can retry and fail, where as it could have failed on the first retry.